bpo-31900: Fix localeconv() encoding for LC_NUMERIC

vstinner opened the pull request:
Decode localeconv() numeric fields like thousands_sep from the
LC_NUMERIC encoding, rather than decoding from the LC_CTYPE encoding.

<!-- issue-number: bpo-31900 -->
https://bugs.python.org/issue31900
<!-- /issue-number -->

vstinner committed:
bpo-31900: Fix localeconv() encoding for LC_NUMERIC

* Add _Py_GetLocaleconvNumeric() function: decode decimal_point and
  thousands_sep fields of localeconv() from the LC_NUMERIC encoding,
  rather than decoding from the LC_CTYPE encoding.
* Modify locale.localeconv() and "n" formatter of str.format() (for
  int, float and complex to use _Py_GetLocaleconvNumeric()
  internally.

stratakis commented:
Tested the PR on a system with glibc 2.26.90 where the test was failing, and it successfully passed.

vstinner committed:
Fix clang warnings on FreeBSD

vstinner committed:
Don't advice to only call localeconv() at startup

The setlocale() is only done in a rare use case (LC_NUMERIC and
LC_CTYPE have a different encoding), it's safe to call localeconv()
in all other cases.

vstinner committed:
Don't declare 'loc' variable twice
